src/Entity.php

```diff
@@ -590,6 +595,17 @@ public function save()
         return $this;
     }
 
+    /**
+     * Generates a primary key
+     *
+     * This method should only be executed from save method.
+     * @codeCoverageIgnore no operations
+     */
+    protected function generatePrimaryKey()
```

Comment:
This function should be moved to the interface.

Reply:
I also think it would make more sense to have this function return the new primary key instead of relying on it setting it to the class member.

Reply:
Oh. I didn't know that you can't have protected methods in interfaces. I kinda understand why, but in this case it would be really useful.
TBH under the given circumstances I would probably simply make the method public and declare it in the interface. It makes the interface a bit less useless and helps with static analysis (and thus IDE support).

Re: the return value, you're right I didn't consider combined keys. Makes sense.

Reply:
But when the interface defines a public method I force the devs to create a public method and I guess there are some devs that want to force other devs to save the entity to get a primary key. At least I would not like that someone can call the method without saving the entity (for example when a sequence is increased).

Reply:
It's not super important to me, I just find it weird to use an interface that doesn't actually declare any interface. One alternative might be an abstract class extending the Entity class... but as I said, it's not super important, I am glad I have a clean way to implement the generator, the rest is details.